### DISCUSSION TITLE
using descriptors with pydantic

### DISCUSSION THREAD

**Discussion by locsmith** [pydantic/pydantic#4278] [Question] [answered]
is it possible to use descriptors with pydantic so the following works...

```python
  class LoggedAccess:

      def __set_name__(self, owner, name):
          self.public_name = name
          self.private_name = '_' + name

      def __get__(self, obj, objtype=None):
          value = getattr(obj, self.private_name)
          print('Accessing %r giving %r', self.public_name, value)
          return value

      def __set__(self, obj, value):
          print('Updating %r to %r', self.public_name, value)
          setattr(obj, self.private_name, value)


  class Test(BaseModel):
      test_field: int = LoggedAccess()
```

so logged access should be called when we do:

```python

test_instance = Test(test_field=1)
```

and the above should print 'Updating test to 1'

**Comment by samuelcolvin** [comment, marked as answer]
See https://github.com/samuelcolvin/pydantic/discussions/4262

**Comment by locsmith** [comment]
That really neat and at the same time frustrating! It does mean that  a descriptor has to be part of a validator and they are no longer independent constructs. Could dealing with  descriptors be worked in with how pydantic works with default values (ie if theres a default but it looks like a descriptor it is a descriptor)? Happy to branch and hack if i could work put where to start... assuming its possible...

**Reply by samuelcolvin** [reply]
There's two issues here:
* satisfying static type analysis
* runtime behaviour

Descriptors would be somewhat helpful for satisfying static type chckers, but as described on #4262, support is too patchy to be useful overall.

In terms of runtime behaviour, any future work on this would need to take place in pydantic-core, in rust where "default" behaviour is managed for pydantic V2.

I think the short answer is that it's not worth it, the code dealing with default values is already fairly complicated and on the "host path", I suspect any work in this area will be wasted when you either discover it's too complicated, or I refuse to accept it because of a performance penalty. But if you're interested, the relevant code is [here](https://github.com/samuelcolvin/pydantic-core/blob/0b2a43b8eb298034b0c569eea97b3caef1818b04/src/validators/typed_dict.rs#L232-L235).

**Reply by samuelcolvin** [reply]
By the way, just in case you're not aware, you could already achieve what you have in your first example using [validator functions](https://pydantic-docs.helpmanual.io/usage/validators/)
